src/derivatives/interestrate/swap.rs

```diff
@@ -88,6 +90,51 @@ pub struct InterestRateSwap<'terms> {
     pub yield_term_structure: Option<&'terms mut YieldTermStructure<'terms>>,
 }
 
+impl InterestRateSwap<'_> {
+    fn add_additional(
+        &mut self,
+        zero_rate: f64,
+        valuation_date: NaiveDate,
+        stripped_curves: &mut Vec<StrippedCurve>,
+    ) -> &mut Vec<StrippedCurve> {
+        stripped_curves.push(StrippedCurve {
+            first_settle_date: self.settle_date(valuation_date),
+            date: self.maturity_date(valuation_date),
+            market_rate: self.fixed_leg.coupon,
+            zero_rate,
+            discount: 0f64,
+            hidden_pillar: false,
+            source: self.yts_type(),
+        });
+        stripped_curves
+    }
+
+    fn calculate_npv(
+        &mut self,
+        zero_rate: f64,
+        valuation_date: NaiveDate,
+        stripped_curves: &mut Vec<StrippedCurve>,
+    ) -> f64 {
+        self.add_additional(zero_rate, valuation_date, stripped_curves);
+        self.npv(valuation_date).unwrap()
+    }
+
+    pub fn solve_zero_rate(
+        &mut self,
+        valuation_date: NaiveDate,
+        stripped_curves: &mut Vec<StrippedCurve>,
+    ) -> f64 {
+        let mut convergency = SimpleConvergency {
+            eps: 1e-15f64,
+            max_iter: 30,
+        };
+        // TODO: can't take self
```

Comment:
I find the compiler error super confusing, but I feel like the issue is that you pass `f` by immutable reference. If you just moved it or passed it by mut reference, it might work?

I.e.
```
// either
let root = find_root_brent(0f64, 1f64, f, &mut convergency);
// or
let root = find_root_brent(0f64, 1f64, &mut f, &mut convergency);
```

Reply:
To clarify (I was half asleep when I wrote the above), the error message I was referring to (and what I assumed your `can't take self` comment referred to) is this:

```
expected a closure that implements the `Fn` trait, but this closure only implements `FnMut`
```

This is confusing because `find_root_brent` takes something that implements `FnMut(F) -> F`, so where does the `Fn` come from?

I believe what's happening is explained by [the traits that auto-implement FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html#implementors):

- Any `&F` (immutable reference) implements `FnMut` as long as `F` is `Fn + Sized` - I think this is why the error says your reference should be `Fn`. This isn't really an option for you, as you do need to mutate (although you could refactor the code and use [interior mutability](https://doc.rust-lang.org/reference/interior-mutability.html) to find a way around it).
- Any `&mut F` (mutable reference) where `F` implements `FnMut` is also `FnMut`.
- And of course you can just move the already `FnMut` closure into `find_root_brent`.

Hope this helps.